storage/redis/peer_store.go

```diff
@@ -0,0 +1,803 @@
+					continue
+				}
+
+				if numWant == 0 {
+					break
+				}
+
+				peers = append(peers, decodePeerKey(serializedPeer(pk.([]byte))))
+				numWant--
+			}
+		}
+	}
+
+	APResult := ""
+	for _, pr := range peers {
+		APResult = fmt.Sprintf("%s Peer:[ID: %s, IP: %s(AddressFamily: %s), Port %d]", APResult, pr.ID.String(), pr.IP.String(), IPver, pr.Port)
+	}
+	log.Debug("storage: AnnouncePeers result", log.Fields{
+		"peers": APResult,
+	})
+
+	return
+}
+
+func (ps *peerStore) ScrapeSwarm(ih bittorrent.InfoHash, addressFamily bittorrent.AddressFamily) (resp bittorrent.Scrape) {
+	select {
+	case <-ps.closed:
+		panic("attempted to interact with stopped redis store")
+	default:
+	}
+
+	resp.InfoHash = ih
+	IPver := addressFamily.String()
+	encodedInfoHash := ih.String()
+	encodedLeecherInfoHash := IPver + "_L_" + encodedInfoHash // key
+	encodedSeederInfoHash := IPver + "_S_" + encodedInfoHash  // key
+
+	conn := ps.rb.open()
+	defer conn.Close()
+
+	leechersLen, err := conn.Do("HLEN", encodedLeecherInfoHash)
+	if err != nil {
+		log.Error("storage: Redis HLEN failure", log.Fields{
+			"Hkey":  encodedLeecherInfoHash,
+			"error": err,
+		})
+		return
+	}
+	lLen := leechersLen.(int64)
+
+	seedersLen, err := conn.Do("HLEN", encodedSeederInfoHash)
+	if err != nil {
+		log.Error("storage: Redis HLEN failure", log.Fields{
+			"Hkey":  encodedSeederInfoHash,
+			"error": err,
+		})
+		return
+	}
+	sLen := seedersLen.(int64)
+
+	if lLen == 0 && sLen == 0 {
+		return
+	}
+
+	resp.Incomplete = uint32(lLen)
+	resp.Complete = uint32(sLen)
+
+	return
+}
+
+// collectGarbage deletes all Peers from the PeerStore which are older than the
+// cutoff time.
+//
+// This function must be able to execute while other methods on this interface
+// are being executed in parallel.
+func (ps *peerStore) collectGarbage(cutoff time.Time) error {
+	select {
+	case <-ps.closed:
+		return nil
+	default:
+	}
+
+	shards := [2]string{bittorrent.IPv4.String(), bittorrent.IPv6.String()}
+
+	conn := ps.rb.open()
+	defer conn.Close()
+
+	cutoffUnix := cutoff.UnixNano()
+	start := time.Now()
+
+	for _, shard := range shards {
+		infohashesList, err := conn.Do("HKEYS", shard) // key
+		if err != nil {
+			return err
+		}
+		infohashes := infohashesList.([]interface{})
+
+		for _, ih := range infohashes {
+			ihStr := string(ih.([]byte))
+
+			ihList, err := conn.Do("HGETALL", ihStr) // field
+			if err != nil {
+				return err
+			}
+			conIhList := ihList.([]interface{})
+
+			if len(conIhList) == 0 {
+				_, err := conn.Do("DEL", ihStr)
+				if err != nil {
+					return err
+				}
+				log.Debug("storage: Deleting Redis", log.Fields{"Hkey": ihStr})
+				_, err = conn.Do("HDEL", shard, ihStr)
+				if err != nil {
+					return err
+				}
+				log.Debug("storage: Deleting Redis", log.Fields{
+					"Hkey":   shard,
+					"Hfield": ihStr,
+				})
+				continue
+			}
+
+			var pk serializedPeer
+			for index, ihField := range conIhList {
+				if index%2 != 0 { // value
```

Comment:
`if index%2 == 1` is more readable, maybe also add a comment why this needs to be checked in the first place, i.e. what is returned from HGETALL

Reply:
done